File: planet-examples/src/main/java/org/saar/minecraft/generator/WorldGenerationPipeline.java
package org.saar.minecraft.generator;

import org.saar.minecraft.Chunk;

public interface WorldGenerationPipeline extends WorldGenerator {

    static WorldGenerationPipeline pipe(WorldGenerator generator) {
        return generator::generateChunk;
    }

    default WorldGenerationPipeline then(WorldGenerator generator) {
        return chunk -> {
            generateChunk(chunk);
            generator.generateChunk(chunk);
        };
    }

    @Override
    void generateChunk(Chunk chunk);

}

File: planet-examples/src/main/java/org/saar/minecraft/generator/WorldGenerationPipeline.kt
package org.saar.minecraft.generator

import org.saar.minecraft.Chunk

fun interface WorldGenerationPipeline : WorldGenerator {
    fun then(generator: WorldGenerator) = WorldGenerationPipeline { chunk ->
        generateChunk(chunk)
        generator.generateChunk(chunk)
    }

    override fun generateChunk(chunk: Chunk)

    companion object {
        @JvmStatic
        fun pipe(generator: WorldGenerator): WorldGenerationPipeline {
            return WorldGenerationPipeline { chunk -> generator.generateChunk(chunk) }
        }
    }
}
